File: agents/router.py
from typing import List, Dict, Any


def route_frameworks(extracted_data: Dict[str, Any], user_selections: List[str]) -> List[str]:
    """
    Deterministic routing based on user selections + content analysis.
    Returns list of framework codes to invoke.
    """
    frameworks = []
    
    # Always include user-selected frameworks
    frameworks.extend(user_selections)
    
    # Auto-trigger based on content (if not already selected)
    
    # ICO/DPA - trigger if personal data detected
    if extracted_data.get("has_personal_data") and "ICO" not in frameworks:
        frameworks.append("ICO")
    
    if extracted_data.get("has_personal_data") and "DPA" not in frameworks:
        frameworks.append("DPA")
    
    # EU AI Act - trigger if biometric or high-risk keywords
    high_risk_keywords = [
        "biometric", "facial", "emotion", "credit scoring", 
        "recruitment", "law enforcement", "border control"
    ]
    
    keywords_text = " ".join(extracted_data.get("keywords", [])).lower()
    system_type = extracted_data.get("system_type", "").lower()
    use_case = extracted_data.get("use_case", "").lower()
    combined_text = f"{keywords_text} {system_type} {use_case}"
    
    if any(kw in combined_text for kw in high_risk_keywords):
        if "EU_AI_ACT" not in frameworks:
            frameworks.append("EU_AI_ACT")
    
    # ISO 42001 - trigger for any system (governance always relevant)
    if "ISO_42001" not in frameworks and len(frameworks) > 0:
        frameworks.append("ISO_42001")
    
    return list(set(frameworks))  # Remove duplicates

File: nextjs-app/src/lib/agents/router.ts
import { ExtractedData, FrameworkCode } from '../backend/types';

const HIGH_RISK_KEYWORDS = [
    'biometric',
    'facial',
    'emotion',
    'credit scoring',
    'recruitment',
    'law enforcement',
    'border control',
    'facial recognition',
    'live facial',
    'watchlist',
];

export function routeFrameworks(extractedData: ExtractedData, userSelections: FrameworkCode[]): FrameworkCode[] {
    const frameworks = new Set<FrameworkCode>(userSelections);

    // Auto-trigger ICO/DPA if personal data detected
    if (extractedData.has_personal_data) {
        frameworks.add('ICO');
        frameworks.add('DPA');
    }

    // Auto-trigger EU AI Act for high-risk indicators
    const combinedText = [
        ...extractedData.keywords,
        extractedData.system_type,
        extractedData.use_case,
        extractedData.deployment_context,
    ]
        .join(' ')
        .toLowerCase();

    if (HIGH_RISK_KEYWORDS.some((kw) => combinedText.includes(kw))) {
        frameworks.add('EU_AI_ACT');
    }

    // Biometric data always triggers EU AI Act
    if (extractedData.has_biometric_data) {
        frameworks.add('EU_AI_ACT');
    }

    // ISO 42001 is always relevant for governance if anything is being evaluated
    if (frameworks.size > 0) {
        frameworks.add('ISO_42001');
    }

    return Array.from(frameworks);
}
